content/courses/advanced-ada/parts/data_types/records.rst

```diff
@@ -2918,20 +2928,684 @@ In this example, :ada:`Arr_Sub` and :ada:`Rec_Sub` are unconstrained subtypes.
+
+       procedure Turn_Off (D : in out Device) is
+       begin
+          D := Device_Off;
+       end Turn_Off;
+
+       procedure Turn_On (D : in out Device) is
+       begin
+          D := Device_On;
+       end Turn_On;
+
+       function Read_Info (D : Device)
+                           return Device_Info is
+         (D.Info);
+
+    end Devices;
+
+Let's create a test application called :ada:`Show_Device` that makes use of
+this device by turning it on and off, and by retrieving information from it:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+    :class: ada-run-expect-failure
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D : Device;
+       I : Device_Info;
+    begin
+       Turn_On (D);
+       I := Read_Info (D);
+
+       Turn_Off (D);
+
+       --  The following call raises
+       --  an exception at runtime
+       --  because D is turned off.
+       I := Read_Info (D);
+    end Show_Device;
+
+In this example, by using the variant part, we're preventing that information
+retrieved by an inappropriate call to the :ada:`Read_Info` function is used
+elsewhere in the application. In fact, if the device is turned off, a call to
+:ada:`Read_Info` raises the :ada:`Constraint_Error` exception because the
+:ada:`Info` component isn't accessible. We see that effect in the
+:ada:`Show_Device` procedure: the call to :ada:`Read_Info` *fails* (by raising
+an exception) when the device has just been turned off.
+
+To avoid exceptions at runtime, we must check the device's state before
+calling :ada:`Read_Info`:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D : Device;
+       I : Device_Info;
+    begin
+       Turn_On (D);
+
+       if D.State = On then
+          I := Read_Info (D);
+       end if;
+
+       Turn_Off (D);
+
+       if D.State = On then
+          I := Read_Info (D);
+       end if;
+    end Show_Device;
+
+Now, no exception is being raised anymore, as we only retrieve information from
+the device when it is turned on |mdash| that is, we only call the
+:ada:`Read_Info` function when the :ada:`State` discriminant of the object is
+set to :ada:`On`.
+
+
+Extending record types
+~~~~~~~~~~~~~~~~~~~~~~
+
+We can use variant parts as a means to extend record types. This can be viewed
+as a static approach to implement type extension |mdash| similar to type
+extension via tagged types, but with clear differences.
+
+Let's say we have a sensor, and we implement a package called :ada:`Sensors`
+that interfaces with that sensor:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Sensors
+
+    package Sensors is
+
+       type Sensor is private;
+
+       type Sensor_Info is
+       record
+          Info_1 : Float := 0.0;
+       end record;
+
+       function Read_Info (S : Sensor)
```

Comment:
Read_Info sure looks like a verb, ie a name for a procedure...  :-)  :-) :-)

Reply:
I'm surprised I haven't noticed that earlier on! Now that you're saying it, `Read_Info` sounds indeed awkward... 

Thanks for pointing it out! I'll come up with a better name. :-)